### PR TITLE
Implement receive in streambuffer

### PR DESCRIPTION
**Pull request by htdvisser**

### CODE REVIEW

File: grpc/streambuffer/streambuffer.go [hunk 1]
```diff
@@ -15,51 +15,82 @@ import (
...
 
+// recvMsg receives a message (possibly dropping a message on full buffers)
+func (s *Stream) recvMsg(msg interface{}) {
+	select {
+	case s.recvBuffer <- msg: // normal flow if the channel is not blocked
+	default:
+		s.log.Debug("streambuffer: dropping received message")
+		atomic.AddUint64(&s.dropped, 1)
+		<-s.recvBuffer // drop oldest and try again (if application temporarily unavailable)
+		select {
```
Comment (rvolosatovs) [inline grpc/streambuffer/streambuffer.go]: why not put this in a `for{}` and `return` once `s.recvBuffer <- msg` succeeds and add a select for? That way old messages will get consumed as long, as `msg` send does not succeed.

```
case s.recvBuffer <- msg:
return
default:
select{
   case <-s.recvBuffer:
   // dropping message
   case s.recvBuffer<-msg:
   return
}
this also applies for `sendMsg`

File: grpc/streambuffer/streambuffer.go [hunk 2]
```diff
@@ -87,12 +118,25 @@ func (s *Stream) Run() (err error) {
...
-		err := stream.RecvMsg(&e)
-		s.log.WithError(err).Debug("streambuffer: error from stream.RecvMsg")
-		recvErr <- err
-		close(recvErr)
+		for {
+			var r interface{}
+			if s.recvFunc != nil {
+				r = s.recvFunc()
+			} else {
+				r = new(empty.Empty)
```
Comment (htdvisser) [inline grpc/streambuffer/streambuffer.go]: I think the `empty.Empty` is pretty well known in protobuf, but sure

File: grpc/streambuffer/streambuffer.go (hunk 1 above)
Comment (htdvisser) [inline grpc/streambuffer/streambuffer.go]: Because these funcs should be non-blocking

Comment (rvolosatovs) [inline grpc/streambuffer/streambuffer.go]: okay, but still, why is only 1 oldest message dropped? just drop all the ones, that are still buffered and don't get sent. it's just `default: return` in the sub-select I wrote then

Comment (htdvisser) [inline grpc/streambuffer/streambuffer.go]: Let's take an example with your proposal:

- A client sends status messages to a server every second.
- The streambuffer has a buffer size of 10
- The server is restarted; the client starts buffering the status messages
- The server is back online, the client regains connectivity and continues sending the buffered message to the server

1. If the server was unreachable for 9 seconds, the client would have buffered 9 messages, and after reconnect, send 9 messages to the server.
2. If the server was unreachable for 11 seconds, the client would have dropped 10 messages, and after reconnect, send only 1 message to the server.

File: grpc/streambuffer/streambuffer.go [hunk 3]
```diff
@@ -126,7 +139,7 @@ func (s *Stream) Run() (err error) {
 			if s.recvFunc != nil {
 				r = s.recvFunc()
 			} else {
-				r = new(empty.Empty)
+				r = new(empty.Empty) // Generic proto message if not interested in received values
```
Comment (rvolosatovs) [inline grpc/streambuffer/streambuffer.go]: what I meant was documenting that `Run()` unmarshals the received message into `empty.Empty`

File: grpc/streambuffer/streambuffer.go [hunk 4]
```diff
@@ -93,6 +103,9 @@ func (s *Stream) recvMsg(msg interface{}) {
 
 // Run the stream
```
Comment (rvolosatovs) [inline grpc/streambuffer/streambuffer.go]: This really doesn't say anything about what this function is for and how it should be used

File: grpc/streambuffer/streambuffer.go [hunk 5]
```diff
@@ -15,51 +15,82 @@ import (
...
 		sendBuffer: make(chan interface{}, bufferSize),
+		recvBuffer: make(chan interface{}, bufferSize),
 		log:        ttnlog.Get(),
 	}
 }
 
-// Stream client->server streaming rpc that buffers (at most) the last {bufferSize} messages
+// Stream implements a buffered streaming RPC
 type Stream struct {
 	// BEGIN sync/atomic aligned
```
Comment (rvolosatovs) [inline grpc/streambuffer/streambuffer.go]: @htdvisser

Comment (htdvisser) [inline grpc/streambuffer/streambuffer.go]: You're commenting on an outdated commit here

File: grpc/streambuffer/streambuffer.go (hunk 3 above)
Comment (htdvisser) [inline grpc/streambuffer/streambuffer.go]: If `recvFunc` is not set, the value is completely ignored, so I don't think that is relevant

File: grpc/streambuffer/streambuffer.go (hunk 5 above)
Comment (rvolosatovs) [inline grpc/streambuffer/streambuffer.go]: it still there and you did not explain

Comment (htdvisser) [inline grpc/streambuffer/streambuffer.go]: screen shot 2017-07-11 at 14 29 09

**Pull request merged**